### PR TITLE
Sbt support

### PR DESCRIPTION
**Pull request by ziyel**
+ Refactor code of create project
    + Use ModuleBuilder instead ProjectTemplateFactory
    + Delete unused files
+ Add sbt option at first page:
image
+ Add sbt version choose at scala project setup step:
image
+ Auto resolve sbt dependencies:
image
image

### CODE REVIEW

**Comment by ziyel** [comment]
For issue https://github.com/Microsoft/azure-tools-for-java/issues/288

**Comment by ziyel** [comment]
Another small change in this PR:
The tips for install scala plugin have been moved to first page of create new hdi project. The tips will show up while user choose the scala hdi project.
scala_plugin

**Comment by voidfunction** [comment]
@testforstephen @adashen Please help review the code due to build script changing.

**Comment by voidfunction** [comment]
>         }

add build tool information here.

---
Refers to: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/HDInsightModuleBuilder.java:178 in 25dbd9a. [](commit_id = 25dbd9af9cd1bf0d4122d241cc05ec405cb2ae5e, deletion_comment = False)

File: BuildDevint [hunk 1]
```diff
@@ -43,7 +43,7 @@ chmod +x ./tools/IntellijVersionHelper
 if [ $INTELLIJ_VERSION == "true" ] ; then
     ./tools/IntellijVersionHelper 2016.3
 fi
-./gradlew buildPlugin --project-dir ./PluginsAndFeatures/azure-toolkit-for-intellij -s -Pintellij_version=IU-2016.3
+./gradlew buildPlugin --project-dir ./PluginsAndFeatures/azure-toolkit-for-intellij -s -Pintellij_version=IU-2016.3 -Pdep_plugins=org.intellij.scala:2016.3.5
```
Comment (adashen) [inline BuildDevint]: Please make sure it works on both 2017/2016

Comment (ziyel) [inline BuildDevint]: Thanks Wei. I will test it.

File: PluginsAndFeatures/azure-toolkit-for-intellij/gradle.properties [hunk 2]
```diff
@@ -1,3 +1,4 @@
 javaVersion=1.8
 sources=false
-intellij_version=IC-2017.1
\ No newline at end of file
+intellij_version=IC-2017.1
+dep_plugins=org.intellij.scala:2017.1.15
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/gradle.properties]: Can we use maven dependency to resolve scala plugin build dependency?

Refer this: http://www.jetbrains.org/intellij/sdk/docs/plugin_repository/api/maven_interface.html

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/HDInsightModuleBuilder.java [hunk 3]
```diff
@@ -128,78 +94,79 @@ public ModuleType getModuleType() {
...
+        createDefaultArtifact(module);
+        switch(this.selectedExternalSystem) {
+            case MAVEN:
+                new MavenProjectGenerator(module, this.selectedTemplate.getTemplateType(), sparkVersion).generate();
+                break;
+            case SBT:
+                new SbtProjectGenerator(module, this.selectedTemplate.getTemplateType(), sparkVersion, sbtVersion).generate();
+                break;
+            default:
+                new MavenProjectGenerator(module, this.selectedTemplate.getTemplateType(), sparkVersion).generate();
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/HDInsightModuleBuilder.java]: Suppose to throw an Unsupported Exception here

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/HDInsightProjectTemplate.java [hunk 4]
```diff
@@ -26,38 +26,41 @@
...
+            case Scala:
+                return "Spark on HDInsight (Scala)";
+            case JavaLocalSample:
+                return "Spark on HDInsight Local Run Sample (Java)";
+            case ScalaLocalSample:
+                return "Spark on HDInsight Local Run Sample (Scala)";
+            case ScalaClusterSample:
+                return "Spark on HDInsight Cluster Run Sample (Scala)";
+            default:
+                return "HDInsight Tools";
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/HDInsightProjectTemplate.java]: Does the type work?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/HDInsightProjectTemplate.java [hunk 5]
```diff
@@ -74,22 +77,31 @@ public String getDescription() {
...
+        switch (this.templateType) {
+            case Java:
+            case JavaLocalSample:
+                return StreamUtil.getImageResourceFile(CommonConst.JavaProjectIconPath);
+            case Scala:
+            case ScalaLocalSample:
+            case ScalaClusterSample:
+                return StreamUtil.getImageResourceFile(CommonConst.ScalaProjectIconPath);
+            default:
+                return StreamUtil.getImageResourceFile(CommonConst.JavaProjectIconPath);
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/HDInsightProjectTemplate.java]: May be null here.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ScalaPluginStatus.java [hunk 6]
```diff
@@ -23,6 +23,5 @@
  */
 public enum ScalaPluginStatus {
     NOT_INSTALLED,
-    INSTALLED,
-    NEED_RESTART
+    INSTALLED
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ScalaPluginStatus.java]: Seems boolean type is enough.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ui/HDInsightProjectTypeStep.form [hunk 7]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ui/HDInsightProjectTypeStep.form]: Good try! 👍

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ScalaPluginStatus.java (hunk 6 above)
Comment (ziyel) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ScalaPluginStatus.java]: Yes. Boolean is enough. Fix it.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ui/HDInsightProjectTypeStep.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+import com.microsoft.azure.hdinsight.projects.HDInsightModuleBuilder;
+import com.microsoft.azure.hdinsight.projects.HDInsightProjectTemplate;
+
+import javax.swing.*;
+import java.awt.event.MouseAdapter;
+import java.awt.event.MouseEvent;
+import java.util.HashSet;
+import java.util.Set;
+
+public class HDInsightProjectTypeStep extends ModuleWizardStep implements Disposable {
```
Comment (voidfunction) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ui/HDInsightProjectTypeStep.java]: >ModuleWizardStep [](start = 46, length = 16)

disposeUIResources() shoule been overrided to release the UI resources

Comment (ziyel) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/projects/ui/HDInsightProjectTypeStep.java]: Fixed.

**Pull request merged**